File: dotnet/bank-simulation/Features/Admin/AdminController.cs
using bank_simulation.Features.Admin.Dto;
using bank_simulation.Shared.Dto;
using Microsoft.AspNetCore.Mvc;

namespace bank_simulation.Features.Admin;

[ApiController]
[Route("api/v1/[controller]")]
public class AdminController : ControllerBase
{
    private readonly IAdminService _adminService;
    public AdminController(IAdminService adminService)
    {
        _adminService = adminService;
    }

    [HttpPost("top-up")]
    public async Task<IActionResult> TopUpCustomerBalance([FromBody] RequestTopUpCustomer request)
    {
        var result = await _adminService.TopUpCustomerBalanceAsync(request);
        return Ok(result);
    }

    [HttpGet("customer")]
    public async Task<IActionResult> ViewCustomers([FromQuery] RequestPaginated request)
    {
        var result = await _adminService.ViewCustomersAsync(request);
        return Ok(result);
    }

    [HttpGet("transaction")]
    public async Task<IActionResult> ViewTransactionCustomers([FromQuery] RequestPaginated request)
    {
        var result = await _adminService.ViewTransactionCustomersAsync(request);
        return Ok(result);
    }
}

File: java/bank_simulation/src/main/java/com/nursyah/bank_simulation/features/admin/AdminController.java
package com.nursyah.bank_simulation.features.admin;

import com.nursyah.bank_simulation.features.admin.dto.RequestTopUpCustomer;
import com.nursyah.bank_simulation.shared.dto.RequestPaginated;
import com.nursyah.bank_simulation.shared.dto.ResponseDto;
import com.nursyah.bank_simulation.shared.dto.ResponsePaginated;
import com.nursyah.bank_simulation.shared.dto.ResponseTransaction;
import com.nursyah.bank_simulation.shared.dto.ResponseUser;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import java.util.concurrent.CompletableFuture;

@RestController
@RequestMapping("/api/v1/admin")
public class AdminController {

    private final IAdminService adminService;

    @Autowired
    public AdminController(IAdminService adminService) {
        this.adminService = adminService;
    }

    @PostMapping("/top-up")
    public CompletableFuture<ResponseEntity<ResponseDto>> topUpCustomerBalance(
            @RequestBody @Validated RequestTopUpCustomer request) {
        return adminService.topUpCustomerBalanceAsync(request)
                .thenApply(ResponseEntity::ok);
    }

    @GetMapping("/customer")
    public CompletableFuture<ResponseEntity<ResponsePaginated<ResponseUser>>> viewCustomers(
            @ModelAttribute RequestPaginated request) {
        return adminService.viewCustomersAsync(request)
                .thenApply(ResponseEntity::ok);
    }

    @GetMapping("/transaction")
    public CompletableFuture<ResponseEntity<ResponsePaginated<ResponseTransaction>>> viewTransactionCustomers(
            @ModelAttribute RequestPaginated request) {
        return adminService.viewTransactionCustomersAsync(request)
                .thenApply(ResponseEntity::ok);
    }
}